keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,1783 @@
+and control plane components.
+
+#### Update sandbox-level QoS-class resources
+
+This future step would be a second extension to the CRI API.
+
+Currently there is no endpoint in the CRI API to update the configuration of
+pod sandboxes. In contrast, container-level resources can be updated with the
+UpdateContainerResources API endpoint. In order to make container and pod
+(sandbox) level QoS-class resources symmetric we want to make it possible to
+update of pod-level resource assignments, too.
+
+This will likely require a new API endpoint in CRI:
+
+```diff
+@@ -38,6 +38,8 @@ service RuntimeService {
+     // RunPodSandbox creates and starts a pod-level sandbox. Runtimes must ensure
+     // the sandbox is in the ready state on success.
+     rpc RunPodSandbox(RunPodSandboxRequest) returns (RunPodSandboxResponse) {}
++    // UpdatePodSandboxConfig updates the configuration of an existing pod sandbox.
++    rpc UpdatePodSandboxConfig(UpdatePodSandboxConfigŔequest) returns (UpdatePodSandboxConfigŔesponse) {}
+```
+
+#### In-place pod vertical scaling
+
+Properly integrating QoS-class resources to
+[In-place pod vertical scaling](#1287) will allow update of running containers.
+However, some QoS-class resources may be immutable, meaning the in-place
+updates are not possible.
+
+#### Access control
+
+This future step adds support for controlling the access to available QoS-class
+resources.
+
+Access control could be achieved e.g. by extending ResourceQuotaSpec which
+would implement restrictions based on the namespace.
+
+```diff
+ // ResourceQuotaSpec defines the desired hard limits to enforce for Quota.
+ type ResourceQuotaSpec struct {
+     // hard is the set of desired hard limits for each named resource.
+     Hard ResourceList
+     // A collection of filters that must match each object tracked by a quota.
+     // If not specified, the quota matches all objects.
+     Scopes []ResourceQuotaScope 
+     // scopeSelector is also a collection of filters like scopes that must match each
+     // object tracked by a quota but expressed using ScopeSelectorOperator in combination
+     // with possible values.
+     ScopeSelector *ScopeSelector
++    // QoSResources contains the desired set of allowed class resources.
++    // +featureGate=QoSResources
++    // +optional
++    QoSResources QoSResourceQuota
+ }
+
++// QoSResourceQuota contains the allowed class resources.
++type QoSResourceQuota struct {
++       // Pod contains the allowed class resources for pods.
++       // +optional
++       Pod []AllowedQoSResource
++       // Container contains the allowed class resources for pods.
++       // +optional
++       Container []AllowedQoSResource
++}
+
++// AllowedQoSResource specifies access to one QoS-class resources type.
++type AllowedQoSResource struct {
++       // Name of the resource.
++       Name QoSResourceName
++       // Allowed classes.
++       Classes []string
++       // Capacity is the hard limit for usage of the class.
```

Comment:
> Do you think this is a count of pods or some scalar based on another resource (e.g. you can only put X many cores into cache class "gold") ?

I thought about counting how many pods (or containers) are using a specific class.

> I am concerned that quota will be the really hard part of this proposal to implement - it's not obvious that a simple count of pods/containers is always meaningful.

I agree that if we want to list other resources under the QoS resource quota it get's a lot hairier. I think it would then probably become something like quota-inside-quota like:

```yaml
spec:
  qosResources:
    container:
      - name: rdt
        classes:
          - name: gold
            hard:
              pods: 2
              containers: 4
              cpu: 4
```

Reply:
blech.  I am not sure what the right answer is, but let's please save some of our "cleverness budget" for this.

Reply:
Sure, it's under future work anyway so not targeting the first implementation phase

Reply:
I don't love that this is punted with no real idea how we might solve it

Reply:
@thockin adding "Capacity" was one of the early feedbacks that we received from @derekwaynecarr. Initially all resources were "infinity", so it can be attached. Derek's concern was "if all are top-priority qos assigned containers, so nobody is top priority", which in some scenarios makes sense. That's why some simple entity counter based (pod/container) capacity was added.

Reply:
ping @thockin